### PR TITLE
Hierarchical NavigationView AutomationPeer Left Mode

### PR DESCRIPTION
**Pull request by ojhad**
Updated NavigationViewItem AutomationPeer to support Expand/Collapse
[Project Board Task](https://github.com/Microsoft/microsoft-ui-xaml/projects/9#card-17909427)

### CODE REVIEW

File: dev/NavigationView/NavigationViewItem.cpp [hunk 1]
```diff
@@ -483,3 +484,16 @@ bool NavigationViewItem::HasChildren()
             MenuItemsSource() ||
             HasUnrealizedChildren());
 }
+
+void NavigationViewItem::RaiseExpandCollapseAutomationEvent(bool isExpanded)
+{
+    if (winrt::AutomationPeer::ListenerExists(winrt::AutomationEvents::PropertyChanged))
+    {
+        auto expandState = isExpanded ? winrt::ExpandCollapseState::Expanded : winrt::ExpandCollapseState::Collapsed;
```
Comment (StephenLPeters) [inline dev/NavigationView/NavigationViewItem.cpp]: >auto expandState = isExpanded ? winrt::ExpandCollapseState::Expanded : winrt::ExpandCollapseState::Collapsed; [](start = 8, length = 109)

move this inside the if block below.

File: dev/NavigationView/NavigationViewItemAutomationPeer.cpp [hunk 2]
```diff
@@ -114,6 +115,16 @@ int32_t NavigationViewItemAutomationPeer::GetSizeOfSetCore()
     return sizeOfSet;
 }
 
+int32_t NavigationViewItemAutomationPeer::GetLevelCore()
+{
+    auto navigationViewItem = Owner().try_as<winrt::NavigationViewItem>();
+    if (navigationViewItem)
```
Comment (StephenLPeters) [inline dev/NavigationView/NavigationViewItemAutomationPeer.cpp]: combine these lines:

if(auto&& navigationViewItem = Owner().try_as<winrt::NavigationViewItem>())

File: dev/NavigationView/NavigationViewItemAutomationPeer.cpp [hunk 3]
```diff
@@ -195,66 +206,70 @@ NavigationViewListPosition NavigationViewItemAutomationPeer::GetNavigationViewLi
...
+        auto listView = winrt::get_self<NavigationView>(navview)->LeftNavListView();
+        auto navigationViewItem = Owner().try_as<winrt::NavigationViewItem>();
+        if (listView && navigationViewItem)
         {
-            if (auto parent = Navigate(winrt::AutomationNavigationDirection::Parent).try_as<winrt::AutomationPeer>())
+            if (auto navListView = listView.try_as<winrt::NavigationViewList>())
             {
-                if (auto children = parent.GetChildren())
+                auto targetNode = winrt::get_self<NavigationViewList>(navListView)->NodeFromContainer(navigationViewItem);
+                if (targetNode)
```
Comment (StephenLPeters) [inline dev/NavigationView/NavigationViewItemAutomationPeer.cpp]: Combine these

File: dev/NavigationView/NavigationViewItemAutomationPeer.h [hunk 4]
```diff
@@ -23,10 +23,17 @@ class NavigationViewItemAutomationPeer :
...
     int32_t GetPositionInSetCore();
     int32_t GetSizeOfSetCore();
+    int32_t GetLevelCore();
+    void RaiseExpandCollapseAutomationEvent(winrt::ExpandCollapseState newState);
 
     // IInvokeProvider
     void Invoke();
 
+    // IExpandCollapseProvider
+    winrt::ExpandCollapseState ExpandCollapseState();
```
Comment (licanhua) [inline dev/NavigationView/NavigationViewItemAutomationPeer.h]: >ExpandCollapseState [](start = 11, length = 19)

It should be in NavigationViewItemDataAutomationPeer just like what we did on treeview.

File: dev/NavigationView/NavigationViewItemAutomationPeer.cpp [hunk 5]
```diff
@@ -195,66 +206,70 @@ NavigationViewListPosition NavigationViewItemAutomationPeer::GetNavigationViewLi
...
                 {
-                    int index = 0;
-                    bool itemFound = false;
-
-                    for (auto const& child : children)
+                    auto targetParentNode = targetNode.Parent();
+                    if (targetParentNode)
```
Comment (StephenLPeters) [inline dev/NavigationView/NavigationViewItemAutomationPeer.cpp]: Combine

File: dev/NavigationView/NavigationViewItemAutomationPeer.cpp [hunk 6]
```diff
@@ -195,66 +206,70 @@ NavigationViewListPosition NavigationViewItemAutomationPeer::GetNavigationViewLi
...
 // In case of calculating the position, if this is the NavigationViewItemAutomationPeer we're iterating through we break the loop.
 int32_t NavigationViewItemAutomationPeer::GetPositionOrSetCountInLeftNavHelper(AutomationOutput automationOutput)
 {
     int returnValue = 0;
-    
+
     if (auto navview = GetParentNavigationView())
     {
-        if (auto listview = winrt::get_self<NavigationView>(navview)->LeftNavListView())
+        auto listView = winrt::get_self<NavigationView>(navview)->LeftNavListView();
```
Comment (StephenLPeters) [inline dev/NavigationView/NavigationViewItemAutomationPeer.cpp]: >auto listView = winrt::get_self<NavigationView>(navview)->LeftNavListView(); [](start = 8, length = 76)

I'd make all these locals const as well

File: dev/NavigationView/NavigationViewItem.cpp [hunk 7]
```diff
@@ -420,12 +420,13 @@ void NavigationViewItem::OnLostFocus(winrt::RoutedEventArgs const& e)
     }
 }
 
-void NavigationViewItem::UpdateIsExpanded(winrt::TreeViewNode node)
+void NavigationViewItem::UpdateIsExpandedForNode(winrt::TreeViewNode node)
```
Comment (StephenLPeters) [inline dev/NavigationView/NavigationViewItem.cpp]: Could pass this variable in a gsl::strict_not_null<winrt::TreeViewNode> to build the not null requirement into the method signature.

**Comment by jevansaks** [comment]
Closing this PR for now -- we'll come back to this when @ojhad is back on WinUI work. :)

**Pull request closed without merging**